File: app/routers/expo_tokens.py
from fastapi import APIRouter, HTTPException, status
from typing import List
from bson import ObjectId
from datetime import datetime

from app.models import (
    ExpoTokenCreate,
    ExpoTokenUpdate,
    ExpoTokenResponse,
    ExpoTokenInDB,
    MessageResponse
)
from app.database import get_expo_tokens_collection

router = APIRouter(prefix="/expo-tokens", tags=["Expo Tokens"])


def token_helper(token: dict) -> ExpoTokenResponse:
    """Convert MongoDB document to ExpoTokenResponse."""
    return ExpoTokenResponse(
        id=str(token["_id"]),
        user_id=token["user_id"],
        expo_token=token["expo_token"],
        created_at=token.get("created_at", datetime.utcnow()),
        updated_at=token.get("updated_at", datetime.utcnow())
    )


@router.post("/", response_model=ExpoTokenResponse, status_code=status.HTTP_201_CREATED)
async def create_expo_token(token_data: ExpoTokenCreate):
    """
    Create a new Expo token entry.
    
    If a token already exists for the user_id, it will be updated instead.
    """
    collection = get_expo_tokens_collection()
    
    # Check if user already has a token
    existing = await collection.find_one({"user_id": token_data.user_id})
    
    if existing:
        # Update existing token
        now = datetime.utcnow()
        await collection.update_one(
            {"user_id": token_data.user_id},
            {
                "$set": {
                    "expo_token": token_data.expo_token,
                    "updated_at": now
                }
            }
        )
        updated = await collection.find_one({"user_id": token_data.user_id})
        return token_helper(updated)
    
    # Create new token entry
    now = datetime.utcnow()
    token_doc = {
        "user_id": token_data.user_id,
        "expo_token": token_data.expo_token,
        "created_at": now,
        "updated_at": now
    }
    
    result = await collection.insert_one(token_doc)
    created = await collection.find_one({"_id": result.inserted_id})
    
    return token_helper(created)


@router.get("/", response_model=List[ExpoTokenResponse])
async def get_all_expo_tokens():
    """Get all Expo tokens."""
    collection = get_expo_tokens_collection()
    tokens = []
    
    async for token in collection.find():
        tokens.append(token_helper(token))
    
    return tokens


@router.get("/{token_id}", response_model=ExpoTokenResponse)
async def get_expo_token(token_id: str):
    """Get a specific Expo token by ID."""
    collection = get_expo_tokens_collection()
    
    if not ObjectId.is_valid(token_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid token ID format"
        )
    
    token = await collection.find_one({"_id": ObjectId(token_id)})
    
    if not token:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expo token not found"
        )
    
    return token_helper(token)


@router.get("/user/{user_id}", response_model=ExpoTokenResponse)
async def get_expo_token_by_user(user_id: str):
    """Get Expo token by user ID."""
    collection = get_expo_tokens_collection()
    
    token = await collection.find_one({"user_id": user_id})
    
    if not token:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No Expo token found for user: {user_id}"
        )
    
    return token_helper(token)


@router.put("/{token_id}", response_model=ExpoTokenResponse)
async def update_expo_token(token_id: str, token_data: ExpoTokenUpdate):
    """Update an Expo token by ID."""
    collection = get_expo_tokens_collection()
    
    if not ObjectId.is_valid(token_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid token ID format"
        )
    
    # Check if token exists
    existing = await collection.find_one({"_id": ObjectId(token_id)})
    if not existing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expo token not found"
        )
    
    # Build update data
    update_data = {k: v for k, v in token_data.model_dump().items() if v is not None}
    
    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No fields to update"
        )
    
    update_data["updated_at"] = datetime.utcnow()
    
    await collection.update_one(
        {"_id": ObjectId(token_id)},
        {"$set": update_data}
    )
    
    updated = await collection.find_one({"_id": ObjectId(token_id)})
    return token_helper(updated)


@router.delete("/{token_id}", response_model=MessageResponse)
async def delete_expo_token(token_id: str):
    """Delete an Expo token by ID."""
    collection = get_expo_tokens_collection()
    
    if not ObjectId.is_valid(token_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid token ID format"
        )
    
    result = await collection.delete_one({"_id": ObjectId(token_id)})
    
    if result.deleted_count == 0:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Expo token not found"
        )
    
    return MessageResponse(message="Expo token deleted successfully")


@router.delete("/user/{user_id}", response_model=MessageResponse)
async def delete_expo_token_by_user(user_id: str):
    """Delete Expo token by user ID."""
    collection = get_expo_tokens_collection()
    
    result = await collection.delete_one({"user_id": user_id})
    
    if result.deleted_count == 0:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"No Expo token found for user: {user_id}"
        )
    
    return MessageResponse(message=f"Expo token for user {user_id} deleted successfully")

File: src/routes/expoTokens.ts
import { getExpoTokensCollection } from "../models/expoToken";
import type { ExpoToken } from "../models/expoToken";
import { ObjectId } from "mongodb";

type RouteParams = Record<string, string>;

interface RequestBody {
  user_id?: string;
  expo_token?: string;
}

async function parseJsonBody(request: Request): Promise<RequestBody | null> {
  try {
    return (await request.json()) as RequestBody;
  } catch {
    return null;
  }
}

function jsonResponse(data: unknown, status: number = 200): Response {
  return new Response(JSON.stringify(data), {
    status,
    headers: { "Content-Type": "application/json" },
  });
}

function errorResponse(message: string, status: number = 400): Response {
  return jsonResponse({ error: message }, status);
}

type RouteHandler = (req: Request, params: RouteParams) => Promise<Response>;

export const expoTokensRoutes: Record<string, RouteHandler> = {
  "POST /api/expo-tokens": async (req: Request, _params: RouteParams) => {
    const body = await parseJsonBody(req);
    if (!body || !body.user_id || !body.expo_token) {
      return errorResponse("user_id and expo_token are required", 400);
    }

    const collection = getExpoTokensCollection();
    const now = new Date();

    const result = await collection.findOneAndUpdate(
      { user_id: body.user_id },
      {
        $set: {
          expo_token: body.expo_token,
          updated_at: now,
        },
        $setOnInsert: {
          created_at: now,
        },
      },
      { upsert: true, returnDocument: "after" }
    );

    return jsonResponse(result, 201);
  },

  "GET /api/expo-tokens": async (_req: Request, _params: RouteParams) => {
    const collection = getExpoTokensCollection();
    const tokens = await collection.find({}).toArray();
    return jsonResponse(tokens);
  },

  "GET /api/expo-tokens/:id": async (req: Request, params: RouteParams) => {
    const id = params.id;
    if (!id) {
      return errorResponse("ID parameter is required", 400);
    }
    if (!ObjectId.isValid(id)) {
      return errorResponse("Invalid ID format", 400);
    }

    const collection = getExpoTokensCollection();
    const token = await collection.findOne({ _id: new ObjectId(id) });

    if (!token) {
      return errorResponse("Expo token not found", 404);
    }

    return jsonResponse(token);
  },

  "GET /api/expo-tokens/user/:user_id": async (req: Request, params: RouteParams) => {
    const user_id = params.user_id;
    if (!user_id) {
      return errorResponse("user_id parameter is required", 400);
    }
    const collection = getExpoTokensCollection();
    const token = await collection.findOne({ user_id });

    if (!token) {
      return errorResponse("Expo token not found for user", 404);
    }

    return jsonResponse(token);
  },

  "PUT /api/expo-tokens/:id": async (req: Request, params: RouteParams) => {
    const id = params.id;
    if (!id) {
      return errorResponse("ID parameter is required", 400);
    }
    if (!ObjectId.isValid(id)) {
      return errorResponse("Invalid ID format", 400);
    }

    const body = await parseJsonBody(req);
    if (!body || (!body.expo_token && !body.user_id)) {
      return errorResponse("At least one field (expo_token or user_id) is required", 400);
    }

    const collection = getExpoTokensCollection();
    const updateData: Partial<ExpoToken> = {
      updated_at: new Date(),
    };

    if (body.expo_token) {
      updateData.expo_token = body.expo_token;
    }
    if (body.user_id) {
      updateData.user_id = body.user_id;
    }

    const result = await collection.findOneAndUpdate(
      { _id: new ObjectId(id) },
      { $set: updateData },
      { returnDocument: "after" }
    );

    if (!result) {
      return errorResponse("Expo token not found", 404);
    }

    return jsonResponse(result);
  },

  // DELETE /api/expo-tokens/:id - Delete expo token by ID
  "DELETE /api/expo-tokens/:id": async (req: Request, params: RouteParams) => {
    const id = params.id;
    if (!id) {
      return errorResponse("ID parameter is required", 400);
    }
    if (!ObjectId.isValid(id)) {
      return errorResponse("Invalid ID format", 400);
    }

    const collection = getExpoTokensCollection();
    const result = await collection.findOneAndDelete({
      _id: new ObjectId(id),
    });

    if (!result) {
      return errorResponse("Expo token not found", 404);
    }

    return jsonResponse({ message: "Expo token deleted successfully" });
  },

  // DELETE /api/expo-tokens/user/:user_id - Delete expo token by user
  "DELETE /api/expo-tokens/user/:user_id": async (req: Request, params: RouteParams) => {
    const user_id = params.user_id;
    if (!user_id) {
      return errorResponse("user_id parameter is required", 400);
    }
    const collection = getExpoTokensCollection();
    const result = await collection.findOneAndDelete({
      user_id,
    });

    if (!result) {
      return errorResponse("Expo token not found for user", 404);
    }

    return jsonResponse({ message: "Expo token deleted successfully" });
  },
};
